tensorflow_addons/layers/tcn.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import absolute_import, division, print_function
+
+import tensorflow as tf
+from tensorflow.keras.layers import (Activation, BatchNormalization, Conv1D,
+                                     Lambda, SpatialDropout1D, add)
+
+
+@tf.keras.utils.register_keras_serializable(package='Addons')
+class ResidualBlock(tf.keras.layers.Layer):
+    """Defines the residual block for the WaveNet TCN.
+
+    Arguments:
+        dilation_rate (int): The dilation power of 2 we are using
+            for this residual block. Defaults to 1.
+        filters (int): The number of convolutional
+            filters to use in this block. Defaults to 64.
+        kernel_size (int): The size of the convolutional kernel. Defaults
+            to 2.
+        padding (String): The padding used in the convolutional layers,
+            'same' or 'causal'. Defaults to 'same'
+        activation (String): The final activation used
+            in o = Activation(x + F(x)). Defaults to 'relu'
+        dropout_rate (Float): Float between 0 and 1. Fraction
+            of the input units to drop. Defaults to 0.0.
+        kernel_initializer (String): Initializer for the kernel weights
+            matrix (Conv1D). Defaults to 'he_normal'
+        use_batch_norm (bool): Whether to use batch normalization in the
+            residual layers or not. Defaults to False.
+        last_block (bool): Whether or not this block is the last residual
+            block of the network. Defaults to False.
+        kwargs: Any initializers for Layer class.
+
+    Returns:
+        A Residual Blcok.
+    """
+
+    def __init__(self,
+                 dilation_rate=1,
+                 filters=64,
+                 kernel_size=2,
+                 padding='same',
+                 activation='relu',
+                 dropout_rate=0.0,
+                 kernel_initializer='he_normal',
+                 use_batch_norm=False,
+                 last_block=False,
+                 **kwargs):
+
+        self.dilation_rate = dilation_rate
+        self.filters = filters
+        self.kernel_size = kernel_size
+        self.padding = padding
+        self.activation = activation
+        self.dropout_rate = dropout_rate
+        self.use_batch_norm = use_batch_norm
+        self.kernel_initializer = kernel_initializer
+        self.last_block = last_block
+
+        super(ResidualBlock, self).__init__(**kwargs)
+
+    def _add_and_activate_layer(self, layer):
+        """Helper function for building layer.
+
+        Arguments:
+            layer: Appends layer to internal layer list and builds it based on
+                    the current output shape of ResidualBlock.
+                    Updates current output shape.
+        """
+        self.residual_layers.append(layer)
+        self.residual_layers[-1].build(self.res_output_shape)
+        self.res_output_shape = self.residual_layers[-1].compute_output_shape(
+            self.res_output_shape)
+
+    def build(self, input_shape):
```

Comment:
Thanks for the feedback, originally I thought all the layers should be initialize in ```build```, I will update so that layers that do not depend on ```input_shape``` will be initialized in ```__init__```, thanks for the feedback!

Reply:
updated! See the latest commit, resolving this conversation. Thanks @WindQAQ